drivers/ipm/ipm_nrfx_ipc.c

```diff
@@ -0,0 +1,253 @@
+			nrfx_vipm_data.callback[event_idx]
+				(nrfx_vipm_data.callback_ctx[event_idx],
+				 0,
+				 NULL);
+		}
+	}
+}
+
+static int vipm_nrf_max_data_size_get(struct device *dev)
+{
+	return ipm_max_data_size_get(dev);
+}
+
+static u32_t vipm_nrf_max_id_val_get(struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return 0;
+}
+
+#define VIPM_DEVICE_1(_idx)									\
+static int vipm_nrf_##_idx##_init(struct device *dev)						\
+{												\
+	if(!nrfx_vipm_data.ipm_init) {								\
+		gipm_init();									\
+		nrfx_vipm_data.ipm_init = true;							\
+	}											\
+	return 0;										\
+}												\
+												\
+static int vipm_nrf_##_idx##_send(struct device *dev, int wait, u32_t id,			\
+				  const void *data, int size)					\
+{												\
+	if (!IS_ENABLED(CONFIG_IPM_MSG_CH_##_idx##_ENABLE)) {					\
+		LOG_ERR("IPM_" #_idx " is not enabled");					\
+		return -EINVAL;									\
+	}											\
+												\
+	if(!IS_ENABLED(CONFIG_IPM_MSG_CH_##_idx##_TX)) {					\
+		LOG_ERR("IPM_" #_idx " is RX message channel");					\
+		return -EINVAL;									\
+	}											\
+												\
+	if (id > NRFX_IPC_ID_MAX_VALUE) {							\
+		return -EINVAL;									\
+	}											\
+												\
+	if (id != 0) {										\
+		LOG_WRN("Passing message ID to IPM with predefined message ID");		\
+	}											\
+												\
+	if (size > 0) {										\
+		LOG_WRN("nRF driver does not support sending data over IPM");			\
+	}											\
+												\
+	return gipm_send(_idx);									\
+}												\
+												\
+static void vipm_nrf_##_idx##_register_callback(struct device *dev,				\
+					       ipm_callback_t cb, void *context)		\
+{												\
+	if(IS_ENABLED(CONFIG_IPM_MSG_CH_##_idx##_RX)) {						\
+		nrfx_vipm_data.callback[_idx] = cb;						\
+		nrfx_vipm_data.callback_ctx[_idx] = context;					\
+	} else {										\
+		LOG_WRN("Trying to register a callback for TX channel IPM_" #_idx);		\
+	}											\
+}												\
+												\
+static int vipm_nrf_##_idx##_set_enabled(struct device *dev, int enable)			\
+{												\
+	if (!IS_ENABLED(CONFIG_IPM_MSG_CH_##_idx##_RX)) {					\
+		LOG_ERR("IPM_" #_idx " is TX message channel");					\
+		return -EINVAL;									\
+	} else if (enable) {									\
+		irq_enable(DT_INST_0_NORDIC_NRF_IPC_IRQ_0);					\
+		nrfx_ipc_receive_event_enable(_idx);						\
+	} else if (!enable) {									\
+		nrfx_ipc_receive_event_disable(_idx);						\
+	}											\
+	return 0;										\
+}												\
+												\
+static const struct ipm_driver_api vipm_nrf_##_idx##_driver_api = {				\
+	.send = vipm_nrf_##_idx##_send,								\
+	.register_callback = vipm_nrf_##_idx##_register_callback,				\
+	.max_data_size_get = vipm_nrf_max_data_size_get,					\
+	.max_id_val_get = vipm_nrf_max_id_val_get,						\
+	.set_enabled = vipm_nrf_##_idx##_set_enabled						\
+};												\
+												\
+DEVICE_AND_API_INIT(vipm_nrf_##_idx , "IPM_"#_idx,						\
+		    vipm_nrf_##_idx##_init, NULL, NULL,						\
```

Comment:
vipm_nrf_##_idx##_init

We should do as little code generation as possible. Co-incidentally,

vipm_nrf_##_idx##_init is the same for all _idx. So we can have vipm_nrf_n_init, and re-use it for _idx.

Reply:
Fixed